### PR TITLE
DrivAerML: cosine T_max=50 with champion EMA+gc stack

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM champion uses cosine T_max=30 (restarts every 30 epochs). On AirfRANS, T_max=50 was dramatically better than T_max=10 — the longer cosine cycle lets EMA track the optimization trajectory more harmoniously, improving both surface and volume metrics simultaneously (+19.4% surface improvement from T_max=10→50). On DM, the T_max=30 optimum was found during early exploration with a very different config (no EMA, no gc). The full champion stack (EMA=0.9995 + gc=0.5) may benefit from a longer cycle.

**The case for T_max=50 on DM:**
- Champion val ceiling is ~3.83% across seeds — this ceiling may be a property of the T_max=30 cosine landscape, not a fundamental limit
- Longer cycles = fewer but deeper restarts = EMA better tracks the slow descent without being disrupted by frequent LR peaks
- We have T_max=24 (nami #3297) and T_max=36 (chrome #3298) in flight, but T_max=50 represents a qualitatively different regime (only ~10 full cycles in 500 epochs vs ~17 at T_max=30)
- The DM training epochs have been extended to 800+ with 600-min budgets — longer T_max harmonizes better with these long training horizons

## Instructions

Run the DM champion config with T_max=50 (instead of 30), keeping everything else identical:

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 50 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --save-checkpoint \
  --eval-interval 10 \
  --seed 42 \
  --wandb-group dm-tmax-sweep \
  --kill-if-best-val-above 150:8.0 \
  --kill-if-no-improvement 250:0.05
```

**CRITICAL:** Use `--no-compile-model` — torch.compile + EMA bug causes frozen val metrics.

**After training**, full-eval on best checkpoint:
```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --load-checkpoint <path-to-best-checkpoint> \
  --eval-only \
  --optimizer adamw --lr 5e-4 --cosine-t-max 50 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --ema-mode fixed --no-compile-model
```

## Baseline

- **DM best val:** 3.833% at ep511 (PR #3072, T_max=30, W&B: ncl1dh88)
- **DM authoritative full-eval TEST:** 4.324% (faye run zx5syby1)
- **External SOTA:** <3.71% (AB-UPT)
- **AF T_max evidence:** T_max=50 at lr=6e-4 gave 0.000482 surface_mse — 19.4% better than T_max=10 (PR #2951)
- **DM T_max sweep (in flight):** T_max=24 (nami #3297), T_max=36 (chrome #3298)

## Expected output

Report:
1. Best val `surface_rel_l2_pct` and epoch
2. W&B run ID
3. Convergence comparison: does T_max=50 reach 4.0% val faster or slower than champion T_max=30?
4. Full-eval TEST from best checkpoint
5. Any stability differences (divergence, instability at cosine peaks?)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Training Run — T_max=50 champion config

| Metric | T_max=50 (ours) | T_max=30 baseline (#3072) | Delta |
|--------|-----------------|---------------------------|-------|
| **Best val `surface_rel_l2_pct`** | **3.914%** (ep990) | **3.833%** (ep511) | +0.081 pp (+2.1%) ❌ |
| **Full-eval TEST** | **4.212%** | **4.324%** (faye zx5syby1) | -0.112 pp (-2.6%) ✓ |
| **Checkpoint-eval TEST** | **4.212%** | **4.685%** (#3072) | -0.473 pp (-10.1%) ✓ |
| **External SOTA (AB-UPT)** | — | <3.71% | gap: 0.204 pp |

- **W&B training run:** `maf97m58` (emma/dm-tmax50-champion)
- **W&B full-eval run:** `xhe9kuud` (emma/dm-tmax50-fulleval)
- **Best epoch:** 990 (of 999 — still improving at termination)
- **Seed:** 42

### Convergence comparison

T_max=50 converges **significantly slower** than T_max=30:

| Milestone | T_max=50 | T_max=30 (est.) |
|-----------|----------|------------------|
| Below 8.0% | ep80 | ~ep80 |
| Below 5.0% | ep190 | ~ep180 |
| Below 4.5% | ep290 | ~ep300 |
| Below 4.0% | **ep790** | **~ep470** |
| Best val | 3.914% @ ep990 | 3.833% @ ep511 |

The longer cosine cycle means fewer restarts (only ~20 full cycles in 999 epochs vs ~33 at T_max=30). The final convergence phase (4.0%→3.9%) took 200 epochs, with the model still descending at every cosine trough. The last four trough readings: ep920=3.948%, ep940=3.936%, ep950=3.923%, ep990=3.914%.

### Stability analysis

- **No divergence or instability** — training was rock-solid throughout 999 epochs
- **Gradient clipping:** 342 of 999 epochs had clip events, declining from 380/394 batches (ep1) to 1/394 (ep985) — very smooth stability profile
- **No NaN or gradient explosions** — gc=0.5 + EMA=0.9995 remains a robust stability guard at T_max=50

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --ema-mode fixed --no-compile-model --save-checkpoint \
  --eval-interval 10 --seed 42 --wandb-group dm-tmax-sweep \
  --kill-if-best-val-above 150:8.0 --kill-if-no-improvement 250:0.05 \
  --agent emma --wandb-name "emma/dm-tmax50-champion"
```

### What happened

**T_max=50 did NOT beat the T_max=30 champion on validation** (3.914% vs 3.833%, +2.1% worse). The hypothesis that a longer cosine cycle would let EMA track the optimization trajectory more harmoniously and break through the 3.83% ceiling was **not confirmed**.

However, there are two notable findings:

1. **Better generalization:** Test metric (4.212%) is significantly better than the champion's authoritative test (4.324%), a 2.6% relative improvement. The val-test gap is only 0.30 pp (vs 0.49 pp for T_max=30), suggesting T_max=50 finds a more generalizable basin.

2. **Still descending at ep999:** The model was clearly still improving at termination — the last 4 cosine troughs each set new bests (3.948→3.936→3.923→3.914). With more epochs, T_max=50 could potentially match or beat 3.833%, but T_max=30 reached that level ~480 epochs earlier. Wall-clock efficiency strongly favors T_max=30.

The 3.83% ceiling appears to be a fundamental landscape property for this config — T_max=50 approaches it from a different trajectory but converges to a similar region. The AirfRANS finding (T_max=50 >> T_max=10) does not transfer to DrivAerML, likely because DM's larger dataset and surface-only task have different optimization dynamics.

### Suggested follow-ups

- **T_max=50 extended run (1500+ epochs):** The model was still improving — an extended budget could test whether T_max=50 eventually matches 3.833%. However, this may not be worth the GPU time given T_max=30 reaches it in half the epochs.
- **T_max=40:** The gap between T_max=30 (3.833%) and T_max=50 (3.914%) suggests the optimum is at or below T_max=30. T_max=40 could bracket more tightly, but the existing T_max=24 (#3297) and T_max=36 (#3298) sweep results are more informative.
- **T_max=50 with higher LR:** Longer cycles could benefit from slightly higher LR to compensate for the slower convergence rate.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3315 because ADVISOR: Closing — val=3.914% with T_max=50, worse than T_max=36 (3.622%) and T_max=30 (3.833%). T_max=50 converges too slowly — model was still at 3.91% after 999 epochs. Schedule optimum is firmly at T_max=36. Full-eval TEST=4.212% is interesting but val is the primary metric..
